File: frontend/src/stores/sources.js
import { defineStore } from "pinia";
import { ref } from "vue";
import sourceService from "@/services/source.service";
import { useToast } from "@/composables/useToast";

export const useSourceStore = defineStore("sources", () => {
  const toast = useToast();

  const items = ref([]);
  const loading = ref(false);

  const fetchList = async () => {
    loading.value = true;
    try {
      const res = await sourceService.list();
      items.value = res.data.data;
    } catch (err) {
      toast.error(err.response?.data?.detail || "Gagal memuat sumber data");
    } finally {
      loading.value = false;
    }
  };

  const create = async (payload) => {
    const res = await sourceService.create(payload);
    toast.success("Sumber data berhasil ditambahkan");
    return res.data.data;
  };

  const update = async (id, payload) => {
    const res = await sourceService.update(id, payload);
    toast.success("Sumber data berhasil diperbarui");
    return res.data.data;
  };

  const remove = async (id) => {
    await sourceService.remove(id);
    toast.success("Sumber data berhasil dihapus");
  };

  return { items, loading, fetchList, create, update, remove };
});

File: backend/app/api/v1/sources.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_current_user
from app.infrastructure.db import get_db
from app.infrastructure.models import User
from app.schemas.common import SuccessResponse
from app.schemas.source import SourceCreateRequest, SourceResponse, SourceUpdateRequest
from app.services.source_service import SourceService

router = APIRouter(prefix="/sources", tags=["sources"])


@router.get("", response_model=SuccessResponse[list[SourceResponse]])
def list_sources(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    sources = SourceService(db).list(current_user.id)
    return SuccessResponse(message="Daftar sumber data", data=[SourceResponse.model_validate(s) for s in sources])


@router.post("", response_model=SuccessResponse[SourceResponse], status_code=status.HTTP_201_CREATED)
def create_source(
    payload: SourceCreateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    source = SourceService(db).create(current_user.id, **payload.model_dump())
    return SuccessResponse(message="Sumber data berhasil ditambahkan", data=SourceResponse.model_validate(source))


@router.put("/{source_id}", response_model=SuccessResponse[SourceResponse])
def update_source(
    source_id: str,
    payload: SourceUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        source = SourceService(db).update(current_user.id, source_id, **payload.model_dump(exclude_unset=True))
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return SuccessResponse(message="Sumber data berhasil diperbarui", data=SourceResponse.model_validate(source))


@router.delete("/{source_id}", response_model=SuccessResponse[None])
def delete_source(source_id: str, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    try:
        SourceService(db).delete(current_user.id, source_id)
    except LookupError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    return SuccessResponse(message="Sumber data berhasil dihapus", data=None)
